Alertmanager segfaults if I don't pass the -alertmanager.web.external-url flag

Not sure if it is an issue. But when I look at the documentation for alertmanager, it doesn't seem to me that `-alertmanager.web.external-url` flag is a required flag. Here is what happens when I don't pass that flag. It works fine with that flag though.


```

$ sudo docker run --rm  --name=alertmanager  --network=host  cortex:v1.0.1 alertmanager.configs.url=http://configs.cortex:3001  -server.http-listen-port 2001 -server.grpc-listen-port 2002 -log.level=debug  -target alertmanager
level=info ts=2020-06-09T15:57:57.010945071Z caller=server.go:147 http=[::]:2001 grpc=[::]:2002 msg="server listening on addresses"
level=debug ts=2020-06-09T15:57:57.012990789Z caller=cluster.go:149 component=cluster msg="resolved peers to following addresses" peers=
level=info ts=2020-06-09T15:57:57.024781121Z caller=cluster.go:161 component=cluster msg="setting advertise address explicitly" addr=10.0.2.15 port=9094
level=debug ts=2020-06-09T15:57:57.0262938Z caller=delegate.go:230 component=cluster received=NotifyJoin node=01EACWTVX1BCH35X2KV4CT1R4D addr=10.0.2.15:9094
level=debug ts=2020-06-09T15:57:57.026771062Z caller=cluster.go:233 component=cluster msg="joined cluster" peers=0
level=info ts=2020-06-09T15:57:57.027893802Z caller=main.go:109 msg="Starting Cortex" version="(version=, branch=, revision=)"
level=info ts=2020-06-09T15:57:57.029518485Z caller=cluster.go:623 component=cluster msg="Waiting for gossip to settle..." interval=200ms
level=debug ts=2020-06-09T15:57:57.029920472Z caller=module_service.go:48 msg="module waiting for initialization" module=alertmanager waiting_for=server
level=info ts=2020-06-09T15:57:57.030184558Z caller=module_service.go:58 msg=initialising module=alertmanager
level=debug ts=2020-06-09T15:57:57.046158906Z caller=multitenant.go:283 component=MultiTenantAlertmanager msg="initial configuration load" num_configs=2
level=debug ts=2020-06-09T15:57:57.046604378Z caller=multitenant.go:312 component=MultiTenantAlertmanager msg="adding configurations" num_configs=2
level=debug ts=2020-06-09T15:57:57.046985744Z caller=multitenant.go:400 component=MultiTenantAlertmanager msg="setting config" user=team-X
level=debug ts=2020-06-09T15:57:57.047624125Z caller=multitenant.go:429 component=MultiTenantAlertmanager msg="initializing new per-tenant alertmanager" user=team-X
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x14da0a0]

goroutine 337 [running]:
github.com/cortexproject/cortex/pkg/alertmanager.New(0xc00017c6e0, 0xc00017c690, 0xc00064f0e0, 0xc00068e300, 0xc)
	/go/src/github.com/cortexproject/cortex/pkg/alertmanager/alertmanager.go:171 +0xe90
github.com/cortexproject/cortex/pkg/alertmanager.(*MultitenantAlertmanager).newAlertmanager(0xc0007ea600, 0xc000654d6a, 0x6, 0xc0006ca770, 0x0, 0x0, 0xc00036c960)
	/go/src/github.com/cortexproject/cortex/pkg/alertmanager/multitenant.go:451 +0x183
github.com/cortexproject/cortex/pkg/alertmanager.(*MultitenantAlertmanager).setConfig(0xc0007ea600, 0xc000654d6a, 0x6, 0xc000122750, 0x8b, 0x0, 0x0, 0x0, 0x2f119a0, 0x3a49680)
	/go/src/github.com/cortexproject/cortex/pkg/alertmanager/multitenant.go:430 +0xac9
github.com/cortexproject/cortex/pkg/alertmanager.(*MultitenantAlertmanager).syncConfigs(0xc0007ea600, 0xc000835140)
	/go/src/github.com/cortexproject/cortex/pkg/alertmanager/multitenant.go:314 +0x2e0
github.com/cortexproject/cortex/pkg/alertmanager.(*MultitenantAlertmanager).starting(0xc0007ea600, 0x3a96880, 0xc00078a780, 0x1, 0xc000836530)
	/go/src/github.com/cortexproject/cortex/pkg/alertmanager/multitenant.go:247 +0x39
github.com/cortexproject/cortex/pkg/util/services.(*BasicService).main(0xc00004c630)
	/go/src/github.com/cortexproject/cortex/pkg/util/services/basic_service.go:134 +0x459
created by github.com/cortexproject/cortex/pkg/util/services.(*BasicService).StartAsync.func1
	/go/src/github.com/cortexproject/cortex/pkg/util/services/basic_service.go:96 +0xa8_
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -20,8 +20,8 @@
 * [BUGFIX] Query-frontend: Fixed rounding for incoming query timestamps, to be 100% Prometheus compatible. #2990
 * [BUGFIX] Querier: query /series from ingesters regardless the `-querier.query-ingesters-within` setting. #3035
 * [BUGFIX] Experimental blocks storage: Ingester is less likely to hit gRPC message size limit when streaming data to queriers. #3015
-* [BUGFIX] Fixes the Alertmanager panicking when no `alertmanager.web.external-url` is provided. #3017
 * [BUGFIX] Fix configuration for TLS server validation, TLS skip verify was hardcoded to true for all TLS configurations and prevented validation of server certificates. #3030
+* [BUGFIX] Fixes the Alertmanager panicking when no `-alertmanager.web.external-url` is provided. #3017
 
 ## 1.3.0 in progress
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -20,6 +20,7 @@
 * [BUGFIX] Query-frontend: Fixed rounding for incoming query timestamps, to be 100% Prometheus compatible. #2990
 * [BUGFIX] Querier: query /series from ingesters regardless the `-querier.query-ingesters-within` setting. #3035
 * [BUGFIX] Experimental blocks storage: Ingester is less likely to hit gRPC message size limit when streaming data to queriers. #3015
+* [BUGFIX] Fixes the Alertmanager panicking when no `alertmanager.web.external-url` is provided. #3017
 * [BUGFIX] Fix configuration for TLS server validation, TLS skip verify was hardcoded to true for all TLS configurations and prevented validation of server certificates. #3030
 
 ## 1.3.0 in progress

--- a/integration/alertmanager_test.go
+++ b/integration/alertmanager_test.go
@@ -106,5 +106,5 @@ func TestAlertmanagerStoreAPI(t *testing.T) {
 	cfg, err = c.GetAlertmanagerConfig(context.Background())
 	require.Error(t, err)
 	require.Nil(t, cfg)
-	require.EqualError(t, err, e2ecortex.ErrNotFound.Error())
+	require.EqualError(t, err, "getting config failed with status 406 and error Alertmanager is not active, revise its configuration\n")
 }

--- a/integration/alertmanager_test.go
+++ b/integration/alertmanager_test.go
@@ -106,5 +106,5 @@ func TestAlertmanagerStoreAPI(t *testing.T) {
 	cfg, err = c.GetAlertmanagerConfig(context.Background())
 	require.Error(t, err)
 	require.Nil(t, cfg)
-	require.EqualError(t, err, "getting config failed with status 406 and error Alertmanager is not active, revise its configuration\n")
+	require.EqualError(t, err, "not found")
 }

--- a/integration/e2ecortex/client.go
+++ b/integration/e2ecortex/client.go
@@ -212,6 +212,10 @@ func (c *Client) GetAlertmanagerConfig(ctx context.Context) (*alertConfig.Config
 		return nil, ErrNotFound
 	}
 
+	if resp.StatusCode/100 != 2 {
+		return nil, fmt.Errorf("getting config failed with status %d and error %v", resp.StatusCode, string(body))
+	}
+
 	var ss *ServerStatus
 	err = json.Unmarshal(body, &ss)
 	if err != nil {

--- a/pkg/alertmanager/multitenant.go
+++ b/pkg/alertmanager/multitenant.go
@@ -484,7 +484,7 @@ func (am *MultitenantAlertmanager) ServeHTTP(w http.ResponseWriter, req *http.Re
 	am.alertmanagersMtx.Unlock()
 
 	if !ok {
-		http.Error(w, "no Alertmanager for this user ID, Alertmanager not configured", http.StatusNotFound)
+		http.Error(w, "the Alertmanager has not been configured", http.StatusNotFound)
 		return
 	}
 

--- a/pkg/alertmanager/multitenant.go
+++ b/pkg/alertmanager/multitenant.go
@@ -175,6 +175,10 @@ func NewMultitenantAlertmanager(cfg *MultitenantAlertmanagerConfig, logger log.L
 		return nil, fmt.Errorf("unable to create Alertmanager data directory %q: %s", cfg.DataDir, err)
 	}
 
+	if cfg.ExternalURL.URL == nil {
+		return nil, fmt.Errorf("unable to create Alertmanager, no 'web.external-url' configured")
+	}
+
 	var fallbackConfig []byte
 	if cfg.FallbackConfigFile != "" {
 		fallbackConfig, err = ioutil.ReadFile(cfg.FallbackConfigFile)

--- a/pkg/alertmanager/multitenant.go
+++ b/pkg/alertmanager/multitenant.go
@@ -483,13 +483,8 @@ func (am *MultitenantAlertmanager) ServeHTTP(w http.ResponseWriter, req *http.Re
 	userAM, ok := am.alertmanagers[userID]
 	am.alertmanagersMtx.Unlock()
 
-	if !ok {
-		http.Error(w, "the Alertmanager has not been configured", http.StatusNotFound)
-		return
-	}
-
-	if !userAM.IsActive() {
-		http.Error(w, "Alertmanager is not active, revise its configuration", http.StatusNotAcceptable)
+	if !ok || !userAM.IsActive() {
+		http.Error(w, "the Alertmanager is not configured", http.StatusNotFound)
 		return
 	}
 

--- a/pkg/alertmanager/multitenant.go
+++ b/pkg/alertmanager/multitenant.go
@@ -483,10 +483,16 @@ func (am *MultitenantAlertmanager) ServeHTTP(w http.ResponseWriter, req *http.Re
 	userAM, ok := am.alertmanagers[userID]
 	am.alertmanagersMtx.Unlock()
 
-	if !ok || !userAM.IsActive() {
-		http.Error(w, "no Alertmanager for this user ID", http.StatusNotFound)
+	if !ok {
+		http.Error(w, "no Alertmanager for this user ID, Alertmanager not configured", http.StatusNotFound)
 		return
 	}
+
+	if !userAM.IsActive() {
+		http.Error(w, "Alertmanager is not active, revise its configuration", http.StatusNotAcceptable)
+		return
+	}
+
 	userAM.mux.ServeHTTP(w, req)
 }
 

--- a/pkg/alertmanager/multitenant.go
+++ b/pkg/alertmanager/multitenant.go
@@ -176,7 +176,7 @@ func NewMultitenantAlertmanager(cfg *MultitenantAlertmanagerConfig, logger log.L
 	}
 
 	if cfg.ExternalURL.URL == nil {
-		return nil, fmt.Errorf("unable to create Alertmanager, no 'web.external-url' configured")
+		return nil, fmt.Errorf("unable to create Alertmanager because the external URL has not been configured")
 	}
 
 	var fallbackConfig []byte

--- a/pkg/alertmanager/multitenant_test.go
+++ b/pkg/alertmanager/multitenant_test.go
@@ -191,7 +191,7 @@ func TestAlertmanager_NoExternalURL(t *testing.T) {
 		DataDir: tempDir,
 	}, log.NewNopLogger(), reg)
 
-	require.EqualError(t, err, "unable to create Alertmanager, no 'web.external-url' configured")
+	require.EqualError(t, err, "unable to create Alertmanager because the external URL has not been configured")
 }
 
 func TestAlertmanager_ServeHTTP(t *testing.T) {
@@ -223,7 +223,7 @@ func TestAlertmanager_ServeHTTP(t *testing.T) {
 
 	resp := w.Result()
 	body, _ := ioutil.ReadAll(resp.Body)
-	require.Equal(t, "no Alertmanager for this user ID, Alertmanager not configured\n", string(body))
+	require.Equal(t, "the Alertmanager is not configured\n", string(body))
 
 	// Create a configuration for the user in storage.
 	mockStore.configs["user1"] = alerts.AlertConfigDesc{
@@ -243,5 +243,5 @@ func TestAlertmanager_ServeHTTP(t *testing.T) {
 
 	resp = w.Result()
 	body, _ = ioutil.ReadAll(resp.Body)
-	require.Equal(t, "Alertmanager is not active, revise its configuration\n", string(body))
+	require.Equal(t, "the Alertmanager is not configured\n", string(body))
 }

--- a/pkg/alertmanager/multitenant_test.go
+++ b/pkg/alertmanager/multitenant_test.go
@@ -7,6 +7,7 @@ import (
 	"context"
 	"fmt"
 	"io/ioutil"
+	"net/http/httptest"
 	"os"
 	"testing"
 
@@ -15,6 +16,7 @@ import (
 	"github.com/prometheus/client_golang/prometheus/testutil"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+	"github.com/weaveworks/common/user"
 
 	"github.com/cortexproject/cortex/pkg/alertmanager/alerts"
 	"github.com/cortexproject/cortex/pkg/util/flagext"
@@ -190,4 +192,56 @@ func TestAlertmanager_NoExternalURL(t *testing.T) {
 	}, log.NewNopLogger(), reg)
 
 	require.EqualError(t, err, "unable to create Alertmanager, no 'web.external-url' configured")
+}
+
+func TestAlertmanager_ServeHTTP(t *testing.T) {
+	mockStore := &mockAlertStore{
+		configs: map[string]alerts.AlertConfigDesc{},
+	}
+
+	externalURL := flagext.URLValue{}
+	err := externalURL.Set("http://localhost:8080/alertmanager")
+	require.NoError(t, err)
+
+	tempDir, err := ioutil.TempDir(os.TempDir(), "alertmanager")
+	require.NoError(t, err)
+	defer os.RemoveAll(tempDir)
+
+	// Create the Multitenant Alertmanager.
+	reg := prometheus.NewPedanticRegistry()
+	am := createMultitenantAlertmanager(&MultitenantAlertmanagerConfig{
+		ExternalURL: externalURL,
+		DataDir:     tempDir,
+	}, nil, nil, mockStore, log.NewNopLogger(), reg)
+
+	// Request when no user configuration is present.
+	req := httptest.NewRequest("GET", externalURL.String(), nil)
+	req.Header.Add(user.OrgIDHeaderName, "user1")
+	w := httptest.NewRecorder()
+
+	am.ServeHTTP(w, req)
+
+	resp := w.Result()
+	body, _ := ioutil.ReadAll(resp.Body)
+	require.Equal(t, "no Alertmanager for this user ID, Alertmanager not configured\n", string(body))
+
+	// Create a configuration for the user in storage.
+	mockStore.configs["user1"] = alerts.AlertConfigDesc{
+		User:      "user1",
+		RawConfig: simpleConfigTwo,
+		Templates: []*alerts.TemplateDesc{},
+	}
+
+	// Make the alertmanager pick it up, then pause it.
+	err = am.updateConfigs()
+	require.NoError(t, err)
+	am.alertmanagers["user1"].Pause()
+
+	// Request when user configuration is paused.
+	w = httptest.NewRecorder()
+	am.ServeHTTP(w, req)
+
+	resp = w.Result()
+	body, _ = ioutil.ReadAll(resp.Body)
+	require.Equal(t, "Alertmanager is not active, revise its configuration\n", string(body))
 }

--- a/pkg/alertmanager/multitenant_test.go
+++ b/pkg/alertmanager/multitenant_test.go
@@ -177,3 +177,17 @@ func TestLoadAllConfigs(t *testing.T) {
 		cortex_alertmanager_config_invalid{user="user3"} 0
 	`), "cortex_alertmanager_config_invalid"))
 }
+
+func TestAlertmanager_NoExternalURL(t *testing.T) {
+	tempDir, err := ioutil.TempDir(os.TempDir(), "alertmanager")
+	require.NoError(t, err)
+	defer os.RemoveAll(tempDir)
+
+	// Create the Multitenant Alertmanager.
+	reg := prometheus.NewPedanticRegistry()
+	_, err = NewMultitenantAlertmanager(&MultitenantAlertmanagerConfig{
+		DataDir: tempDir,
+	}, log.NewNopLogger(), reg)
+
+	require.EqualError(t, err, "unable to create Alertmanager, no 'web.external-url' configured")
+}

